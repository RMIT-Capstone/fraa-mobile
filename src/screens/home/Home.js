import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { object } from 'prop-types';
import { View, Text, Image } from 'react-native';
import styles from './HomeStyle';
import { getAttendanceSessionsState } from '../../redux/reducers/AttendanceSessionsReducer';
const CheckInIcon = require('../../assets/CheckInIcon.png');

const Home = () => {
  // hard code here to avoid restarting app every time code changes to see the effect
  const [sessions, setSessions] = useState([
    {
      lecturer: 'Yossi Nygate',
      location: 'SGS/2.4.041 (Lab-Mac)',
      courseName: 'Capstone Project - Part A',
      createdAt: '2020-10-27T12:55:00.209Z',
      validOn: '2020-10-28T08:42:00.577Z',
      expireOn: '2020-10-28T08:43:00.577Z',
      courseCode: 'OENG1183',
      id: 'A3eYst4EtyXpSA6RQF98',
    },
    {
      lecturer: 'Yossi Nygate',
      location: 'SGS/2.4.041 (Lab-Mac)',
      courseName: 'Capstone Project - Part B',
      createdAt: '2020-10-27T12:55:00.209Z',
      validOn: '2020-10-28T08:44:00.577Z',
      expireOn: '2020-10-28T08:45:00.577Z',
      courseCode: 'OENG1183',
      id: 'A3eYst4EtyXpSA6RQF98',
    },
  ]);
  const [displaySession, setDisplaySession] = useState({});
  const [isHappening, setIsHappening] = useState(false);
  const [timeDifference, setTimeDifference] = useState({ hours: '', minutes: '' });

  const loadDisplaySession = () => {
    sessions.forEach((session) => {
      const { expireOn, validOn } = displaySession;
      let rightNow = new Date();

      if (session) {
        setIsHappening(rightNow > new Date(validOn) && rightNow < new Date(expireOn));
      }

      if (rightNow > new Date(expireOn)) {
        let array = sessions.slice(1);
        setSessions(array);
        setDisplaySession(array[0]);
      }
    });
  };

  useEffect(() => {
    setDisplaySession(sessions[0]);
    loadDisplaySession();

    let interval = null;

    interval = setInterval(() => {
      loadDisplaySession();
    }, 1000);

    return () => {
      clearInterval(interval);
    };
  });

  useEffect(() => {
    let interval = null;

    if (sessions.length !== 0 && !isHappening) {
      interval = setInterval(() => {
        let now = new Date();
        let timeDifference1 = new Date(displaySession.validOn) - now;
        let truncated = Math.trunc(timeDifference1 / 1000);
        setTimeDifference({ hours: Math.floor(truncated / 3600), minutes: Math.floor((truncated % 3600) / 60) + 1 });
      }, 1000);
    }

    return () => {
      clearInterval(interval);
    };
  });

  const transformSessionTime = (time) => {
    const timeObj = new Date(time);
    return timeObj.toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', hour12: true });
  };

  if (sessions.length === 0) {
    return (
      <View style={[styles.container, styles.centered]}>
        <Text>No events today. Take a break, get some rest.</Text>
      </View>
    );
  } else {
    const { validOn, courseName, location } = displaySession;
    const today = new Date();
    const validOnDateObject = new Date(validOn);
    const dateOfValidOn = validOnDateObject.getDate();
    const monthOfValidOn = validOnDateObject.toLocaleDateString(undefined, { month: 'long' });
    const dayOfValidOn = validOnDateObject.toLocaleDateString(undefined, { weekday: 'long' });

    return (
      <View style={[styles.container, styles.centered]}>
        <View style={[styles.topChildContainer, styles.centered]}>
          <View
            style={[styles.timeIndicatorContainer, isHappening ? styles.happening : styles.upcoming, styles.centered]}>
            <Text style={styles.timeIndicatorText}>{isHappening ? 'Happening' : 'Upcoming'}</Text>
          </View>
          <View style={[styles.eventContainer, styles.raised]}>
            <View style={[styles.firstEventRow]}>
              <Text style={styles.time}>{transformSessionTime(validOn)}</Text>
              <Text style={styles.courseName}>{courseName}</Text>
              <Text style={styles.location}>{location}</Text>
            </View>
            <View style={[styles.secondEventRow]}>
              <Text style={styles.day}>{dateOfValidOn === today.getDate() ? 'Today' : dayOfValidOn}</Text>
              <Text style={styles.date}>{dateOfValidOn}</Text>
              <Text style={styles.month}>{monthOfValidOn}</Text>
            </View>
          </View>
          {isHappening ? (
            <View style={[styles.checkInBtnContainer, styles.activeBtn, styles.raised, styles.centeredRow]}>
              <Image source={CheckInIcon} style={styles.checkInIcon} />
              <Text style={styles.checkInText}>Check In</Text>
            </View>
          ) : (
            <View style={[styles.checkInBtnContainer, styles.disabledBtn, styles.raised, styles.centered]}>
              <Text style={styles.disabledText}>
                {timeDifference.hours} hours and {timeDifference.minutes} minutes
              </Text>
              <Text style={styles.disabledText}>before session</Text>
            </View>
          )}
        </View>
        <View style={[styles.bottomChildContainer, styles.centered]}>
          <View style={[styles.infoContainer, styles.raised, styles.centered]}>
            <Text style={styles.infoText}>You have missed 3 sessions for this course</Text>
          </View>
        </View>
      </View>
    );
  }
};

Home.propTypes = {
  navigation: object.isRequired,
  attendanceSessions: object.isRequired,
};

const mapStateToProps = (state) => ({
  attendanceSessions: getAttendanceSessionsState(state),
});

export default connect(mapStateToProps, null)(Home);
