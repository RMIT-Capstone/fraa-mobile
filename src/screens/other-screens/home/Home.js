import React from 'react';
import { arrayOf, object, bool, string } from 'prop-types';
import { View, Text, TouchableOpacity, TouchableWithoutFeedback, Image } from 'react-native';
import styles from './HomeStyle';
import { navigateTo } from '../../../helpers/navigation';
import ROUTES from '../../../navigation/routes';

const CheckInIcon = require('../../../assets/CheckInIcon.png');

const Home = ({
  email,
  homeScreenSessions,
  isLoadingSessions,
  displaySession,
  isHappening,
  timeDifference,
  navigation,
  registeredLocally,
  locationPermission,
  tooFar,
}) => {
  const noEvents = () => {
    if (Array.isArray(homeScreenSessions) && homeScreenSessions.length === 0) {
      return true;
    }
    return typeof displaySession === 'object' && displaySession === {};
  };

  const transformSessionTime = (time) => {
    const timeObj = new Date(time);
    return timeObj.toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', hour12: true });
  };

  if (isLoadingSessions) {
    return (
      <View style={[styles.container, styles.centered]}>
        <Text>Loading sessions...</Text>
      </View>
    );
  }

  if (noEvents()) {
    return (
      <View style={[styles.container, styles.centered]}>
        <Text>No events today. Take a break, get some rest.</Text>
      </View>
    );
  }

  const {
    id,
    validOn,
    room,
    course: { courseName },
    attendees,
  } = displaySession;

  const checkedIn = () => attendees.includes(email);

  const renderCheckInButton = () => {
    if (checkedIn()) {
      return (
        <TouchableOpacity disabled style={[styles.checkedInBtn, styles.disabledBtn, styles.raised, styles.centered]}>
          <Text style={styles.disabledText}>Checked In!</Text>
        </TouchableOpacity>
      );
    }
    if (isHappening) {
      return (
        <TouchableOpacity
          onPress={() => navigateTo(navigation, ROUTES.CAMERA, { fromHome: false, id })}
          style={[styles.checkInBtnContainer, styles.activeBtn, styles.raised, styles.centeredRow]}>
          <Image source={CheckInIcon} style={styles.checkInIcon} />
          <Text style={styles.checkInText}>Check In</Text>
        </TouchableOpacity>
      );
    }

    if (!isHappening) {
      return (
        <View style={[styles.checkInBtnContainer, styles.disabledBtn, styles.raised, styles.centered]}>
          <Text style={styles.disabledText}>
            {timeDifference.hours} hours and {timeDifference.minutes} minutes
          </Text>
          <Text style={styles.disabledText}>before session</Text>
        </View>
      );
    }

    if (!locationPermission) {
      return <Text style={styles.disabledText}>Please allow location services and restart FRAA</Text>;
    }
    if (tooFar) {
      return (
        <TouchableWithoutFeedback
          style={[styles.checkInBtnContainer, styles.disabledBtn, styles.raised, styles.centered]}>
          <Text style={styles.disabledText}>Please come to the classroom to check-in</Text>
        </TouchableWithoutFeedback>
      );
    }
    if (!registeredLocally) {
      return <Text style={styles.disabledText}>You need to register your identity in Profile</Text>;
    }

    return null;
  };

  const today = new Date();
  const validOnDateObject = new Date(validOn);
  const dateOfValidOn = validOnDateObject.getDate();
  const monthOfValidOn = validOnDateObject.toLocaleDateString(undefined, { month: 'long' });
  const dayOfValidOn = validOnDateObject.toLocaleDateString(undefined, { weekday: 'long' });

  return (
    <View style={[styles.container, styles.centered]}>
      <View style={[styles.topChildContainer, styles.centered]}>
        <View
          style={[styles.timeIndicatorContainer, isHappening ? styles.happening : styles.upcoming, styles.centered]}>
          <Text style={styles.timeIndicatorText}>{isHappening ? 'Happening' : 'Upcoming'}</Text>
        </View>
        <View style={[styles.eventContainer, styles.raised]}>
          <View style={[styles.firstEventRow]}>
            <Text style={styles.time}>{transformSessionTime(validOn)}</Text>
            <Text style={styles.courseName}>{courseName}</Text>
            <Text style={styles.location}>{room}</Text>
          </View>
          <View style={[styles.secondEventRow]}>
            <Text style={styles.day}>{dateOfValidOn === today.getDate() ? 'Today' : dayOfValidOn}</Text>
            <Text style={styles.date}>{dateOfValidOn}</Text>
            <Text style={styles.month}>{monthOfValidOn}</Text>
          </View>
        </View>
        {renderCheckInButton()}
      </View>
      <View style={[styles.bottomChildContainer, styles.centered]}>
        <View style={[styles.infoContainer, styles.raised, styles.centered]}>
          <Text style={styles.infoText}>You have missed 3 sessions for this course</Text>
        </View>
      </View>
    </View>
  );
};

Home.propTypes = {
  email: string.isRequired,
  homeScreenSessions: arrayOf(object).isRequired,
  isLoadingSessions: bool.isRequired,
  displaySession: object,
  isHappening: bool.isRequired,
  timeDifference: object.isRequired,
  navigation: object.isRequired,
  registeredLocally: bool.isRequired,
  tooFar: bool.isRequired,
  locationPermission: bool.isRequired,
};

Home.defaultProps = {
  displaySession: {},
};

export default Home;
