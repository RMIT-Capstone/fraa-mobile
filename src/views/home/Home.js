// @flow
import React from 'react';
import {object} from 'prop-types';
import {View, Text, Button} from 'react-native';
import styles from './HomeStyle';
import {navigateTo} from '../../helpers/navigation';
import ROUTES from '../../routes/constants';
type Props = {
  navigation: object,
};
const Home = ({navigation}: Props) => {
  return (
    <View testID="Home" style={styles.centered}>
      <Text style={styles.text}>This is home</Text>
      <Button
        testID="Home_testBtn"
        // title="Go to Test"
        // onPress={() => navigateTo(navigation, ROUTES.TEST)}
        title="Open Camera"
        onPress={() => navigateTo(navigation, ROUTES.CHECK_IN)}
      />
    </View>
  );
};

Home.propTypes = {
  navigation: object.isRequired,
};

export default Home;
