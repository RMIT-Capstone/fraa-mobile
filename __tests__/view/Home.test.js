// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import Home from '../../src/views/home/Home';

test('renders Home correctly', () => {
  const navigation = {};
  const tree = renderer.create(<Home navigation={navigation} />).toJSON();
  expect(tree).toMatchSnapshot();
});
