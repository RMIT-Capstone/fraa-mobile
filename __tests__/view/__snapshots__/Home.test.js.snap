// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Home correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="Home"
>
  <Text
    style={
      Object {
        "color": "#000054",
      }
    }
  >
    This is home
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={Object {}}
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="Home_testBtn"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Open Camera
      </Text>
    </View>
  </View>
</View>
`;
